src/main/java/org/reaktivity/nukleus/http2/internal/stream/Http2ServerFactory.java

```diff
@@ -2187,16 +2245,21 @@ private void onDecodeData(
                 {
                     exchange.cleanup(traceId, authorization);
                     doEncodeRstStream(traceId, authorization, streamId, error);
+                    progress += payloadRemaining.value;
                 }
                 else
                 {
-                    if (dataLength > 0)
+                    final int payloadLength = payload.capacity();
+
+                    if (payloadLength > 0)
                     {
-                        final DirectBuffer payload = http2Data.payload();
-                        exchange.doRequestData(traceId, authorization, payload, 0, payload.capacity());
+                        payloadRemaining.set(payloadLength);
+                        exchange.doRequestData(traceId, authorization, payload, payloadRemaining);
+                        progress = payloadLength - payloadRemaining.value;
```

Comment:
This should probably be `+=` for consistency with `+=` below and would remain consistent even if we choose to make `progress` non-zero in future.